[macareonie] iP

macareonie opened the pull request:
# Tsundere Chatbot
>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

This chatbot frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/macareonie/ip/releases/).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you 😉

And it is __FREE!__

Features:

- [x] Managing todos 
- [x] Managing deadlines
- [x] Managing events
- [x] Saved sessions
- [x]  GUI (Coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the method for launching the GUI:


```
    public void start(Stage stage) {
        try {
            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
            AnchorPane ap = fxmlLoader.load();
            Scene scene = new Scene(ap);
            stage.setScene(scene);
            fxmlLoader.<MainWindow>getController().setTsundere(tsun);
            stage.show();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

macareonie committed:
Add Greet and Exit function + Renamed Chatbot

macareonie committed:
Add echo function

macareonie committed:
Add Add task and list function

macareonie committed:
Add Task, Action classes with function to mark tasks as done

macareonie committed:
Add ToDo, Event, Deadline classes to differentiate task types

macareonie committed:
Fixed typo in Event

macareonie committed:
Removed emoticons temporarily due to conflict with test scripts

macareonie committed:
Add test scripts for I/O

macareonie committed:
Add GeneralException and error handling for incorrect user input

macareonie committed:
Add Delete funtion for tasks

macareonie committed:
Fixed bug with delete function + Updated test script for I/O

macareonie committed:
Add Save feature

macareonie committed:
Merge branch 'save-feature'

macareonie committed:
Add LocalDate compatibility for Deadline tasks

Deadline object takes in a String to be set as its deadline.

LocalDate is more sensible because users can read the deadline in a
clear format. Deadline cannot just be any String and will reflect a
proper date.

Let's,
* Change the Deadline object to use LocalDate for deadline
* Update remaining code to accomodate the change

macareonie committed:
Merge branch 'branch-level-8'

macareonie committed:
Refactor closely related code as classes

Source code contains code handling many different aspects in the same
files.

Separating code into their respective aspects is more readable
and allows for tidy code.

Let's,
* Abstract all the code into 4 main classes: Ui, Storage, Parser and
  TaskList

macareonie committed:
Merge branch 'branch-MoreOOP'

macareonie committed:
All files are in the same directory of main

Separating related class files into their own packages increases
ease of reading and understanding the code and their function

Let's,
 * Package related class files together

macareonie committed:
Merge branch 'branch-Packages'

macareonie committed:
Merge branch 'add-gradle-support'

macareonie committed:
Add Gradle Checkstyle functionality

macareonie committed:
Fix issue with running 'run' using Gradle

macareonie committed:
Add JUnit tests for Task, ToDo, Event and Deadline

Ensure classes are free of bugs

Multiple tests to check to correctness of the various methods in each
class so that the code can run smoothly without errors

macareonie committed:
Merge branch 'branch-junit'

macareonie committed:
Update .gitignore

macareonie committed:
Rectify issue with jar file

macareonie committed:
Package App as a JAR file

Note: Running the JAR file creates a separate data folder under
ip/build/lib where it is located. This data folder is different from the
data folder that stores information when the code is run locally.

macareonie committed:
Update code to follow given coding standard

Following a common coding standard allows for readability and ease of
understanding code across everyone who may access the source code

macareonie committed:
Merge branch 'branch-A-CodingStandard'

macareonie committed:
Add JavaDoc to source code

JavaDoc allows for ease of understanding code.

macareonie committed:
Merge branch 'branch-A-JavaDoc'

macareonie committed:
Add Find function

Allows users to filter for tasks with certain keywords.

Let's,
 * Update Parser to include "find" as a command.
 * Update TaskList to include "find" functionality.

macareonie committed:
Merge branch 'branch-Level-9'

macareonie committed:
Add CheckStyle capability using Gradle

macareonie committed:
Merge branch 'A-CheckStyle'

macareonie committed:
Add GUI using JavaFX

Chatbot uses a Command-Line Interface.

A GUI is more pleasing to the eyes and user-friendly as users can
experience a more realistic chatting experience rather than through the
command line.

Let's, configure the necessary fxml files and update existing code to
fit a transtition to a GUI chatbot.

macareonie committed:
Merge branch 'branch-Level-10'

Jerome-j commented on src/main/java/tsundere/Main.java:
```diff
@@ -0,0 +1,32 @@
+    private Tsundere tsun = new Tsundere();
+
+    @Override
+    public void start(Stage stage) {
```
Consider adding a JavaDoc to the `start` method

Jerome-j commented on src/main/java/tsundere/Tsundere.java:
```diff
@@ -0,0 +1,22 @@
+import tsundere.task.Storage;
+
+
+import javafx.application.Application;
```
Consider running the Java check-style.

The imports should be like this:

```java
package tsundere;

import javafx.application.Application;
import tsundere.task.Storage;
```

Jerome-j commented on src/main/java/tsundere/Tsundere.java:
```diff
@@ -0,0 +1,22 @@
+     * Initializes Tsundere with Storage
+     */
+    public Tsundere() {
+        storage = new Storage();
```
Consider referencing `storage` as `this.storage` to prevent ambiguity.

Jerome-j commented on src/main/java/tsundere/Tsundere.java:
```diff
@@ -0,0 +1,22 @@
+
+
+import javafx.application.Application;
+
```
Consider adding JavaDoc to the `Tsundere.java` file

Jerome-j commented on src/main/java/tsundere/gui/MainWindow.java:
```diff
@@ -0,0 +1,90 @@
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            } catch (IOException e) {
+                System.out.println(e);
```
Consider printing `System.out.println(e.getMessage());` instead of what you are currently doing as outputting `e` may unintentionally expose sensitive information, such as the internal information of the exception to the User.

Jerome-j commented on src/main/java/tsundere/task/Storage.java:
```diff
@@ -0,0 +1,120 @@
+            new File("./data").mkdirs();
+            TaskList.taskList = loadTasksFromFile();
+        } catch (IOException | GeneralException e) {
+            System.out.println("Something went wrong with loading your previous session data!");
```
Consider providing this error string as a public static final String CONSTANT.

Jerome-j commented on src/main/java/tsundere/task/Task.java:
```diff
@@ -0,0 +1,61 @@
+
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
Great job on boolean variable naming!

Jerome-j commented on src/main/java/tsundere/task/Task.java:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * Sets Task completion status to true.
+     */
+    public void markAsDone() {
```
Consider renaming the function to `markTaskAsDone()` so that the function becomes even more explicit.

Jerome-j commented on src/main/java/tsundere/task/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+     * @throws GeneralException If TaskList is empty.
+     */
+    public static String unmark() throws GeneralException {
+
```
Please consider using Egyptian brackets as specified in coding standard

```java
        if (TaskList.taskList.isEmpty()) {
            throw new GeneralException("What you tryna unmark huh?")
        }
```

Jerome-j commented on src/main/java/tsundere/task/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+        try {
+            Task t = TaskList.taskList.get(Integer.parseInt(Parser.name.substring(7, 8)) - 1);
+            if (t.getStatusIcon().equals(" ")) {
+                return("You haven't even started this task dummy!");
```
Hahaha I particularly like this remark!

Jerome-j commented on src/main/java/tsundere/task/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+            }
+        } catch (NumberFormatException e) {
+            throw new GeneralException("Can't you spell?");
+        } catch (IndexOutOfBoundsException e ) {
```
There should not be a space as such `e )` but should be `catch (IndexOutOfBoundsException e) {` instead.

Jerome-j commented on src/main/java/tsundere/task/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+        } catch (NumberFormatException e) {
+            throw new GeneralException("Can't you spell?");
+        } catch (IndexOutOfBoundsException e ) {
+            throw new GeneralException("Im pretty sure that's the wrong task number! Check again!");
```
Consider using proper English here, "I am" or "I'm" instead "Im"

Jerome-j commented on src/main/java/tsundere/task/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+     */
+    public static String mark() throws GeneralException {
+
+        if (TaskList.taskList.isEmpty()) throw new GeneralException("What you tryna mark huh?");
```
Please consider using Egyptian brackets here.

Jerome-j commented on src/main/java/tsundere/task/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+        if (TaskList.taskList.isEmpty()) throw new GeneralException("What you tryna mark huh?");
+        try {
+            Task t = TaskList.taskList.get(Integer.parseInt(Parser.name.substring(5, 6)) - 1);
+            if (t.getStatusIcon().equals("X")) {
```
Consider checking based on the boolean value, rather than doing a more "long-winded" way.

Jerome-j commented on src/main/java/tsundere/task/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+            }
+        } catch (NumberFormatException e) {
+            throw new GeneralException("Can't you spell?");
+        } catch (IndexOutOfBoundsException e ) {
```
Same thing with the spacing here.

Jerome-j commented on src/main/java/tsundere/task/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+     */
+    public static String delete() throws GeneralException {
+
+        if (TaskList.taskList.isEmpty()) throw new GeneralException("What you tryna delete huh?");
```
Consider using Egyptian brackets.

Jerome-j commented on src/main/java/tsundere/task/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+            TaskList.taskList.remove(idx);
+            return getListSize("deleted", t);
+        } catch (NumberFormatException e) {
+            throw new GeneralException("Can't you spell?");
```
Consider changing this error message to be a more descriptive one, since the user might have entered a negative integer which is technically not spelling.

Jerome-j commented on src/main/java/tsundere/task/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+    public static String list() throws GeneralException {
+        StringBuilder response = new StringBuilder();
+        int size = TaskList.taskList.size();
+        if (TaskList.taskList.isEmpty()) throw new GeneralException("Aren't you pretty free now? "
```
Please consider using Egyptian bracket.

Jerome-j commented on src/main/java/tsundere/task/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+     *
+     * @throws GeneralException If command given violates given format.
+     */
+    public static String addToDo() throws GeneralException {
```
Consider renaming the function to `addTodoTask()` to be more descriptive.

Jerome-j commented on src/main/java/tsundere/task/TaskList.java:
```diff
@@ -0,0 +1,215 @@
+    public static String find() throws GeneralException {
+
+        int size = TaskList.taskList.size();
+        if (TaskList.taskList.isEmpty()) throw new GeneralException("Theres's nothing to find here!");
```
Consider adding Egyptian bracket.

Jerome-j commented on src/main/java/tsundere/task/ToDo.java:
```diff
@@ -0,0 +1,36 @@
+package tsundere.task;
+
+public class ToDo extends Task {
```
Consider adding a JavaDoc to public class/functions.

Jerome-j commented on src/test/java/tsundere/task/DeadlineTest.java:
```diff
@@ -0,0 +1,40 @@
+    @Test
+    public void testDeadlineToSaveString() {
+        Deadline deadline = new Deadline("test ", "Jan 01 2024");
+        assertEquals("D,0,test ,Jan 01 2024", deadline.toSaveString());
```
Consider leaving the space after the comma: `assertEquals("D, 0, test, Jan 01 2024",`

Jerome-j commented on src/test/java/tsundere/task/DeadlineTest.java:
```diff
@@ -0,0 +1,40 @@
+
+public class DeadlineTest {
+    @Test
+    public void testDeadlineToString() {
```
Consider naming your test case like this: `intDivision_nonZeroDivisor_success ` as specified in https://nus-cs2103-ay2324s2.github.io/website/schedule/week3/topics.html#c-to-java-junit-junit-basic in W3.7d C++ to Java → JUnit → JUnit: Basic

Jerome-j commented on src/test/java/tsundere/task/DeadlineTest.java:
```diff
@@ -0,0 +1,40 @@
+        Deadline deadline = new Deadline("test ", "Jan 01 2024");
+        deadline.markAsDone();
+        deadline.unMark();
+        assertEquals(" ", deadline.getStatusIcon());
```
Consider running `assertEqual` on the whole `deadline` object.

Jerome-j commented on src/test/java/tsundere/task/EventTest.java:
```diff
@@ -0,0 +1,40 @@
+    @Test
+    public void testEventToSaveString() {
+        Event event = new Event("test ", "1 ", "2");
+        assertEquals("E,0,test ,1 ,2", event.toSaveString());
```
Consider writing with the spaces after the comma.

Jerome-j commented on src/main/java/tsundere/task/Task.java:
```diff
@@ -0,0 +1,61 @@
+     * @param description Name of Task.
+     */
+    public Task(String description) {
+        this.description = description;
```
Consider running the `trim()` command so that you remove the trailing and leading whitespace.

Jerome-j commented on src/test/java/tsundere/task/DeadlineTest.java:
```diff
@@ -0,0 +1,40 @@
+    @Test
+    public void testDeadlineUnMark() {
+        Deadline deadline = new Deadline("test ", "Jan 01 2024");
+        deadline.markAsDone();
```
Consider running an assert statement here too.

Jerome-j reviewed (commented):
Hello @macareonie Ryan, I enjoyed interacting with your chatbot. It looks so cool and has a cool personality. In particular, I loved your customisation to the GUI and it was really very refreshing to interact with it.

However, there are somethings that I spotted that you might need to consider adding: JavaDocs to at least 50% of the public methods as specified in `A-JavaDoc` Level.

There are some minor problems with the formatting of the code which I have pointed out. I would maybe recommend you to make use of the A-CheckStyle level where these issues can easily surface and be resolved.

In addition, I think you can consider storing all the "hard-coded" String messages as a constant - so that it will be easily changed if necessary.

It was great learning from you!

Thank you very much.

Have a nice day.

Best regards,
Jerome

macareonie committed:
Add assertions to various code sections.

Assertions exist as an indicator for logically impossible situations
that may arise when running the code. It verifies the assumptions made
about the whole source code.

Let's,
 * Add assertion statements to Task and subclass constructors.
 * Add assertion statements to Storage switch statement default clause.

macareonie committed:
Merge pull request #1 from macareonie/branch-A-Assertions

Add assertions to various code sections.

macareonie committed:
Add Tagging functionality for tasks.

Tasks contain just their general description and duration or deadline if
applicable.

Adding tags provide additional information for each task.

Let's,
 * Add an array that allows for maximum of 3 tags per task.
 * Update TaskList and Parser to accomodate tagging and untagging tasks.
 * Update Storage to allow saving and loading of tasks and their tags.

macareonie committed:
Merge branch 'branch-C-Tags'

macareonie committed:
Merge branch 'branch-C-Tags'

macareonie committed:
Merge branch 'master' of https://github.com/macareonie/ip

macareonie committed:
Refactor code where necessary

Code is untidy and lacking in quality in some places.

Using comments from peer PR reviews as well as examining it personally,
refactoring the code makes it more readable,efficient, clean and prevents
possibilities of future bugs.

Let's,
 * Review the comments from peer reviews.
 * Ensure all code complies with Java coding standard.
 * Add new Javadoc and update existing ones.

macareonie committed:
Merge pull request #2 from macareonie/branch-A-CodeQuality

Refactor code where necessary

macareonie committed:
Fix formatting issues regarding Task details

The formatting for information regarding Deadline and Event objects
specifically is not consistent and included unnecessary leading and
trailing whitespaces.

Let's,
 * Use the trim command to remove such whitespaces when parsing the user
   input command.
 * Update Junit tests to accomodate the new changes.

macareonie committed:
Fix bug with storing tagged Tasks

Tagged tasks and untagged task cannot be saved/loaded into data savefile
due to indexing errors in the relevant Storage methods. Introducing a
new boolean indicating if a Task has tags can help with converting the
Task details into a properly formatted savestring.

Let's,
 * Add a new hasTags boolean in Tasks.
 * Edit Storage load and parse data methods.
 * Update Junit tests.

macareonie committed:
Set up Continuous Integration

Add a gradle.yml file as per instructions that runs an automated build
process to detect integration errors as soon as possible.

macareonie committed:
Ensure code compliance with checkstyle guidelines

Code is unable to pass Continuous Integration automated build checks due
to violations of checkstyle requirements.

Let's,
 * Rectify all checkstyle violations.

macareonie committed:
Fix bug with code due to checkstyle changes

Changing code from a previous commit to comply with checkstyle
guidelines is causing commands to result in NullPointerExceptions
because the commands are not registered in the Parser object.

Let's,
 * Edit TaskList to make use of name variable from Parser.

macareonie committed:
Improve GUI

Current GUI has inconsistent padding between images, text and borders.
The GUI also consists a solid colour background.

Let's,
 * Add colour changes for Tsundere dialogbox and User dialogbox.
 * Implement appropriate padding changes.

macareonie committed:
Merge branch 'branch-A-BetterGui'

macareonie committed:
Add Help command and finetune GUI

Adding a help command allows inexperienced users to find a list of
commands that can be used.

Let's,
 * Add help command functionality.
 * Update checkstyle.
 * Add SVG icon and title to GUI title bar.

macareonie committed:
Add UserGuide

A user guide helps consolidate all information regarding the program for
all parties to read, be it new or experienced users.

macareonie committed:
Merge branch 'branch-A-UserGuide'

macareonie committed:
Add acknowledgements for images used

macareonie committed:
Rename storage file

Renaming storage file to something more specific prevent conflicts
with data filenames used by
other similar programs.

macareonie committed:
Add icons to README.md

macareonie committed:
Adjust GUI dimensions

GUI appears too large on laptop with smaller screen size as compared to
an external monitor. Reducing the height of the GUI will help.

macareonie committed:
Update README.md
